packages/big-design/src/components/Checkbox/Checkbox.tsx

```diff
@@ -18,102 +19,92 @@ interface PrivateProps {
-          isIndeterminate={isIndeterminate}
-          checked={checked}
-          htmlFor={id}
-          aria-hidden={true}
-        >
-          {!checked && isIndeterminate ? <RemoveIcon /> : <CheckIcon />}
-        </StyledCheckbox>
-        {this.renderLabel()}
-      </CheckboxContainer>
-    );
-  }
-
-  private getInputId() {
-    const { id } = this.props;
-
-    return id ? id : this.uniqueId;
-  }
-
-  private renderLabel() {
-    const htmlFor = this.getInputId();
-    const { disabled, hiddenLabel, label } = this.props;
-
+const RawCheckbox: React.FC<CheckboxProps & PrivateProps> = ({
+  checked,
+  className,
+  disabled,
+  hiddenLabel,
+  isIndeterminate,
+  label,
+  forwardedRef,
+  style,
+  ...props
+}) => {
+  const id = useMemo(() => (props.id ? props.id : uniqueId('checkBox_')), [props.id]);
+  const labelId = useMemo(() => uniqueId('checkBox_label_'), []);
+
+  const renderedLabel = useMemo(() => {
     if (typeof label === 'string') {
       return (
-        <StyledLabel
-          disabled={disabled}
-          hidden={hiddenLabel}
-          htmlFor={htmlFor}
-          aria-hidden={disabled}
-          id={this.labelUniqueId}
-        >
+        <CheckboxLabel disabled={disabled} hidden={hiddenLabel} htmlFor={id} aria-hidden={disabled} id={labelId}>
           {label}
-        </StyledLabel>
+        </CheckboxLabel>
       );
     }
 
-    if (React.isValidElement(label) && label.type === Checkbox.Label) {
+    if (React.isValidElement(label) && label.type === CheckboxLabel) {
       return React.cloneElement(label as React.ReactElement<React.LabelHTMLAttributes<HTMLLabelElement>>, {
         hidden: hiddenLabel,
-        htmlFor,
-        id: this.labelUniqueId,
+        htmlFor: id,
+        id: labelId,
       });
     }
 
-    return null;
-  }
-}
+    if (!label) {
+      return null;
+    }
 
-const CheckboxWithForwardedRef = React.forwardRef<HTMLInputElement, CheckboxProps>(
-  ({ className, style, ...props }, ref) => <RawCheckbox {...props} forwardedRef={ref} />,
+    warning('label must be either a string or a CheckboxLabel component.');
+  }, [disabled, hiddenLabel, id, label, labelId]);
+
+  return (
+    <CheckboxContainer className={className} style={style}>
+      <HiddenCheckbox
+        type="checkbox"
+        checked={checked}
+        id={id}
+        disabled={disabled}
+        {...props}
+        aria-labelledby={labelId}
+        ref={checkbox => {
+          if (checkbox && typeof isIndeterminate === 'boolean') {
+            checkbox.indeterminate = !checked && isIndeterminate;
+          }
+
+          if (forwardedRef) {
+            if (typeof forwardedRef === 'function') {
+              forwardedRef(checkbox);
+            } else {
+              // RefObject.current is readonly in DefinitelyTyped but in practice you
+              // can still write to it.
+              // See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31065
+              // @ts-ignore
```

Comment:
🍹 would doing `(forwardRef as any).current = checkbox`  be a bit less destructive?

Reply:
To be more explicit, I could do `(forwardedRef as React.MutableRefObject<HTMLInputElement | null>).current = checkbox;` instead. What do you think?